File: Frontend/emp-management-frontend/src/app/services/employee.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Employee } from '../model/employee';

@Injectable({
  providedIn: 'root'
})
export class EmployeeService {

  baseUrl: string = 'http://localhost:8080/api/v1/employees';
  // baseurlPost: string = 'http://localhost:8080/api/v1/employee';
  employees: Employee[] = [];
  constructor(private http: HttpClient) {
   }

//GET ALL EMPLOYEE
getAllEmployees(): Observable<Employee[]> {
  return this.http.get<Employee[]>(`${this.baseUrl}`)
}

// Create an employee
createEmployee(employee: Employee): Observable<Employee> {
  return this.http.post<Employee>(`${this.baseUrl}` , employee);
}

// GET Employee by Id:
getEmployeeById(employeeId: string): Observable<Employee>{
   return this.http.get<Employee>(`${this.baseUrl}/${employeeId}`);
}

// update employee
updateEmployee(employeeId: string , employee: Employee): Observable<Employee>{
  return this.http.put<Employee>(`${this.baseUrl}/${employeeId}` , employee)
}

// delete employee
deleteEmployee(employeeId: string): Observable<void>{
 return this.http.delete<void>(`${this.baseUrl}/${employeeId}`)
}
}

File: Backend/emp-management/src/main/java/com/employee/management/services/EmployeeService.java
package com.employee.management.services;

import java.util.List;

import org.springframework.stereotype.Service;

import com.employee.management.exceptions.ResourceNotFoundException;
import com.employee.management.model.Employee;
import com.employee.management.repository.EmployeeRepository;

@Service
public class EmployeeService {

	private EmployeeRepository employeeRepository;

	// constructor injection. by default it wires our dependency.
	public EmployeeService(EmployeeRepository employeeRepository) {
		this.employeeRepository = employeeRepository;
	}

	// GET All Employees
	public List<Employee> getAllEmployees() {
		List<Employee> allEmps = this.employeeRepository.findAll();
		if (allEmps.size() <= 0) {
			throw new ResourceNotFoundException(" no employees present");
		}
		return allEmps;
	}

	// GET employee by id
	public Employee getEmployeeById(String id) {
		Employee emp = this.employeeRepository.fetchItemById(id);
		return emp;
	}

	// Create Employee
	public Employee createEmployee(Employee employee) {
		Employee newEmployee = this.employeeRepository.save(employee);
		return newEmployee;
	}

	// Update Employee
	public Employee updateEmployee(String id, Employee employee) {
		Employee previousEmployee = getEmployeeById(id);
		previousEmployee.setEmployeeId(employee.getEmployeeId());
		previousEmployee.setFirstName(employee.getFirstName());
		previousEmployee.setLastName(employee.getLastName());
		previousEmployee.setEmail(employee.getEmail());
		// save the updated previous employee
		Employee updatedEmployee = this.createEmployee(previousEmployee);
		return updatedEmployee;
	}

	// delete employee by id
	public void deleteById(String employeeId) {
		this.employeeRepository.deleteById(employeeId);
	}
}
